airflow/providers/amazon/aws/waiters/base_waiter.py

```diff
@@ -0,0 +1,37 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+from botocore.waiter import Waiter, WaiterModel, create_waiter_with_client
+
+from airflow.providers.amazon.aws.hooks.base_aws import BaseAwsConnection
+
+
+class BaseBotoWaiter:
+    """
+    Used to create custom Boto3 Waiters.
+
+    For more details, see airflow/providers/amazon/aws/waiters/README.md
+    """
+
+    def __init__(self, client: BaseAwsConnection, model_config: dict) -> None:
```

Comment:
`BaseAwsConnection` use mostly for type annotations (which not fully implemented yet) and it is union type of `boro3.client` and `boto3.resource`

`boto3.client` is subclass of `botocore.client.BaseClient` (plus some magic and factories) and waiters if it exists is a part of this client. However `boto3.resource` as far as I know doesn't have waiters access directly but you always could get actual client of resource by call property `resource.meta.client`.

Sooo.. I think annotations as `boto3.client` would be enough.